pkg/trace/agent/agent.go

```diff
@@ -72,6 +73,7 @@ func NewAgent(ctx context.Context, conf *config.AgentConfig) *Agent {
 	agnt := &Agent{
 		Concentrator:       stats.NewConcentrator(conf.BucketInterval.Nanoseconds(), statsChan),
 		Blacklister:        filters.NewBlacklister(conf.Ignore["resource"]),
+		Validator:          filters.NewValidator(conf.RequiredTags),
```

Comment:
Should be good enough to do this:
```go
for _, tag := range conf.RequiredTags {
    if _, ok := root.Meta[tag]; !ok {
		log.Debugf("Trace discarded due to missing required keys. root: %v", root)
		atomic.AddInt64(&ts.TracesFiltered, 1)
		atomic.AddInt64(&ts.SpansFiltered, tracen)
        continue traceLoop
    }
}
```
It can all be done with very little code. No new concepts, no new documentation, no new files, no new struct field. All clear when reading 🙂

Reply:
Thanks for the review. Since we were also supposed to add `RejectedTags`, should I have two separate for loops to check through each of the lists? Or would it be better to have a separate struct to handle that logic?

Reply:
Sorry but I'm lacking a bit of knowledge here. What is `RejectedTags` supposed to do? Is it supposed to reject a trace if the root contains a specific tag? Basically the inverse of `RequiredTags`?

If so, maybe we can just have a method called `func (a *Agent) tagsValidated(root pb.Span) bool` which contains both for loops?

I think a struct field and a new construct should not be needed. What do you think? Perhaps I'm lacking some info 🙂

Reply:
Since the struct doesn't really hold any state to share between runs, I think we could skip it...

I'm not sure what `RejectedTags` does but I'll assume it rejects a trace if the root contains any of those tags? If so, we can just add a new method, something like:
```go
// tagsValidated reports whether the given span passes validation by ensuring that
// it's not missing any of the required tags and has none of the rejected ones.
func (a *Agent) tagsValidated(span pb.Span) bool {
    // the two for loops in here
}
```

Reply:
Something like this:
```
func (a *Agent) Validates(span *pb.Span) error {
	for _, tag := range v.reqTags {
		if _, ok := span.Meta[tag]; !ok {
			return errors.New("required tag(s) missing")
		}
	}
	for _, tag := range v.rejectTags {
		if _, ok := span.Meta[tag]; ok {
			return errors.New("invalid tag(s) found")
		}
	}
	return nil
}
```
?

Reply:
Something like that will work. 

Be sure to have a look at this comment for understanding what feature is required:
https://github.com/DataDog/datadog-agent/issues/7050#issuecomment-765113552

The user needs to reject based not just on key but on value.
i.e. they might have all their services reporting the tag `"app:*something*"`, and they want to reject all spans containing `"app:specific-reject-value"`, not all spans containing the `"app"` key.

Reply:
This seems to be more complex then. I have missed looking at the issue, my bad. It seems like we need to compile regular expressions too. In that case, it makes sense to have a separate struct because compiling the regular expressions every time on every trace will cost too much CPU. A `TagValidator` (not just `Validator` which is too ambiguous) will be needed, so you can expand on your current implementation.

* You will need a structure to pick up the config map, similar to [this one](https://github.com/DataDog/datadog-agent/blob/9b632ddbca35e59b556c91c527e840a4eb054fa5/pkg/trace/config/apply.go#L89-L102).
* You will need to register all new options [here](https://github.com/DataDog/datadog-agent/blob/9b632ddbca35e59b556c91c527e840a4eb054fa5/pkg/config/apm.go#L14).
* You'll have to parse them in [this function](https://github.com/DataDog/datadog-agent/blob/master/pkg/trace/config/apply.go) and have them added as a field to [this structure](https://github.com/DataDog/datadog-agent/blob/9b632ddbca35e59b556c91c527e840a4eb054fa5/pkg/trace/config/config.go#L49).
* Finally, when initializing the `TagValidator`, you'll have to pass the rules (let's call the struct `TagRules`) to it and compile the regular expressions, similar to how we do in [the blacklister](https://github.com/DataDog/datadog-agent/blob/9b632ddbca35e59b556c91c527e840a4eb054fa5/pkg/trace/filters/blacklister.go#L31-L49).

Reply:
I'm not sure if the current proposal is the best. If someone wants us to expand on it later, we might be in trouble. I suggested an alternative. Please check the issue and let me know what you think. Sorry to jump in late on this. If we do go down that route, and have config, etc. a new structure may make sense.